enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,281 @@
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+For clarity in this doc we set the definition for "Control Plane" as "The collection of stateless and stateful processes which enable a Kubernetes cluster to meet minimum operational requirements". This includes: kube-apiserver, kube-controller-manager, kube-scheduler, kubelet and Etcd.
+
+This proposal outlines a solution for declaratively managing as a single entity the compute resources that host the OCP Control Plane components. It introduces scaling and self-healing capabilities for this compute resources while honouring inviolable Etcd expectations and with out disrupting the lifecycle of Control plane components.
+
+## Motivation
+
+The Control Plane is the most critical and sensitive entity of a running cluster. Today OCP Control Plane instances are "pets" and therefore fragile. There are multiple scenarios where adjusting the compute capacity which is backing the Control Plane components might be desirable either for resizing or repairing. 
+Currently there is nothing that automates or eases this task. The steps for the Control Plane to be resized in any manner or to recover from a tolerable failure are completely manual. Different teams are following different SOP documents scattered around with manual steps resulting in lost of information, confusion and extra efforts for engineers and users.
+
+### Goals
+
+- To have a declarative mechanism to manage the Control Plane as a single entity.
+- To support declarative safe horizontal scaling towards an odd number replicas for the control plane compute resources.
+- To support declarative safe vertical scaling for the control plane compute resources.
+- To support declarative safe self-healing and replacement of Control Plane compute resources when the unhealthy resources are below n/2.
+- To support even spread of compute resources across multiple failure domains.
+
+### Non-Goals
+
+- To integrate with any existing Etcd topology e.g external clusters. Stacked Etcd with dynamic member identities, local storage and the Cluster Etcd Operator are an assumed invariant.
+- To managed individual Control Plane components. Self hosted Control Plane components that are self managed by their operators is an assumed invariant.
+- To integrate with any Control Plane components topology e.g Remote hosted pod based Control Plane.
+- Automated disaster recovery of a cluster that has lost quorum.
+- To manage OS upgrades. This is managed by the Machine Config Operator.
+- To manage configuration changes at the software layer. This is managed by the Machine Config Operator.
+- To automate the provisioning and decommission of the bootstrapping instance managed by the installer.
+
+## Proposal
+
+This proposes an ad-hoc CRD and controller for declaratively managing as a single entity the compute resources that host the OCP Control Plane components.
```

Comment:
It's not clear to me from this proposal why we need this component. Why aren't MachineSets and Machine health checks sufficient for managing the compute of a control plane? Am I missing something?

Reply:
I tried to clarify why only machineSets and MHC are not enough here https://github.com/openshift/enhancements/pull/278/commits/0aca4f198a751301df8eb044ae202a76b6b80ead

Reply:
I'm resolving this comment to keep this discussio in a single thread -> https://github.com/openshift/enhancements/pull/278/files#diff-dcd7423e408e8b740b1b9b14e53eeda1R65